# Level 13

This time an image of a [phone](/13/disprop.jpg) is given. It's hard to see households that have a phone that looks like the image given because nowadays everyone uses Iphones, Galaxy etc...


The name of the image is `disprop.jpg`.


The HTML title is call him. 


Below the phone is a text in yellow that says **phone that evil**. 


Who am I supposed to call? then I remembered in level 12 the response was `b'Bert is evil! go back!\n'` for the code below.


```python
import requests

res = requests.get("http://www.pythonchallenge.com/pc/return/evil4.jpg", auth=("huge", "file"))
print(res.content) # b'Bert is evil! go back!\n'
```

So it looks like I need to call `Bert` but I still don't know how to call Bert.


In the page source theres a **[link](http://www.pythonchallenge.com/pc/phonebook.php)** to a PHP page.


In the PHP page there's some XML snippet. When googling it it says it's a fault response from an XML-RPC communication.

```xml 

<methodResponse>
<fault>
    <value>
        <struct>
            <member>
                <name>faultCode</name>
            <value>
                <int>105</int>
            </value>
        </member>
        <member>
        <name>faultString</name>
            <value>
                <string>XML error 5: empty document</string>
            </value>
                </member>
            </struct>
        </value>
    </fault>
</methodResponse>
```

XML-RPC is a way that lets me run functions or procedures that lives on a different computer. It write messages in a special XML format and uses HTTP to send messages.


Python has an **[XML-RPC](/https://docs.python.org/3/library/xmlrpc.html)** module. Personally I haven't seen anyone use it. 


XML-RPC has a **[xmlrpc.client.ServerProxy](https://docs.python.org/3/library/xmlrpc.client.html#xmlrpc.client.ServerProxy)** class that allows you to interact with a remote XML-RPC server as if it were a local Python object.


You can think of `xmlrpc.client.ServerProxy` as a Python tool that lets you talk to a program on another computer as if that program's functions were right there in your own code, handling all the complex communication details for you.


The sad thing is when I try to make a connection using the code below I get a gzip error `raise BadGzipFile('Not a gzipped file (%r)' % magic) gzip.BadGzipFile: Not a gzipped file (b'\n\n')`

```python
import xmlrpc.client

conn = xmlrpc.client.ServerProxy("http://www.pythonchallenge.com/pc/phonebook.php")
print(conn.system.listMethods())
```


I didn't know what to do and was stuck here for a while. I even asked a question on **[stack-overflow](https://stackoverflow.com/questions/79114964/xmlrpc-client-error-on-python-3-13-macos-apple-silicon)**. 


I also answered the question myself below. It took me 2 months to figure the solution out lol. A day after I posted the question on stack-overflow someone else posted the same issues on the **[python-challenge forum](https://groups.google.com/g/python-challenge/c/EavRGvegY1w)**.


The people on the forum even gave me a hint. With the hint I wrote the solution.


```python
class MyTransport(themoduleyouneedtouse.Transport):
    accept_gzip_encoding = False
```


With the hint above I rewrote the code. 


```python
import xmlrpc.client

class MyTransport(xmlrpc.client.Transport):
    accept_gzip_encoding = False 

conn = xmlrpc.client.ServerProxy("http://www.pythonchallenge.com/pc/phonebook.php", transport=MyTransport())

print(conn.system.listMethods()) # ['phone', 'system.listMethods', 'system.methodHelp', 'system.methodSignature', 'system.multicall', 'system.getCapabilities']
```

**[Serverproxy.system.listMethods](https://docs.python.org/3/library/xmlrpc.client.html#xmlrpc.client.ServerProxy.system.listMethods)** returns a list of methods supported.


Hooray it looks like the code isn't having the `gzip` problem like before.

```python
import xmlrpc.client

class MyTransport(xmlrpc.client.Transport):
    accept_gzip_encoding = False 

conn = xmlrpc.client.ServerProxy("http://www.pythonchallenge.com/pc/phonebook.php", transport=MyTransport())

# print(conn.system.listMethods())

print(conn.phone('Bert')) # 555-ITALY
```

Calling `Bert` using the phone method returns `555-ITALY` which seems to be the answer to level14.


When changing the URL to `555-ITALY` I get a 404 Error. According to wikipedia **[555](https://en.wikipedia.org/wiki/555_(telephone_number))** isn't a real number. 
So removing `555` and just changing the URL to **[ITALY](http://www.huge:file@pythonchallenge.com/pc/return/ITALY.html)** might be the real answer. 


On the page it says **SMALL letters.** so **[italy](http://www.pythonchallenge.com/pc/return/italy.html)** must be the answer and it is. 


Until I solved level13 I didn't know phone numbers that started with a `555` are a hoax. Maybe I didn't know it because I'm not american lol.


I personally had a hard-time solving this level. I especially had a hard-time on how to bypass the `gzip` error. 
I still don't 100% understand how someone could look at the traceback output of the code and think of writing a class to not get the `gzip error` also XML-RPC was something I've never heard before so that was very unfamiliar to me. 


```python
File "/home/pythonchallenge/13/13.py", line 15, in <module>
    print(conn.system.listMethods())
          ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.12/xmlrpc/client.py", line 1122, in __call__
    return self.__send(self.__name, args)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.12/xmlrpc/client.py", line 1461, in __request
    response = self.__transport.request(
               ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.12/xmlrpc/client.py", line 1166, in request
    return self.single_request(host, handler, request_body, verbose)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.12/xmlrpc/client.py", line 1182, in single_request
    return self.parse_response(resp)
           ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.12/xmlrpc/client.py", line 1342, in parse_response
    while data := stream.read(1024):
                  ^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.12/gzip.py", line 324, in read
    return self._buffer.read(size)
           ^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.12/_compression.py", line 68, in readinto
    data = self.read(len(byte_view))
           ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.12/gzip.py", line 527, in read
    if not self._read_gzip_header():
           ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.12/gzip.py", line 496, in _read_gzip_header
    last_mtime = _read_gzip_header(self._fp)
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.12/gzip.py", line 456, in _read_gzip_header
    raise BadGzipFile('Not a gzipped file (%r)' % magic)
gzip.BadGzipFile: Not a gzipped file (b'\n\n')
```


Tip of the level would be to check the python-challenge forum when you're stuck and can't do anything... 